fs/btrfs/zstd.c

```diff
@@ -50,10 +51,30 @@ static void zstd_free_workspace(struct list_head *ws)
 	kfree(workspace);
 }
 
-static struct list_head *zstd_alloc_workspace(void)
+static int zstd_set_level(struct list_head *ws, unsigned level)
+{
+	struct workspace *workspace = list_entry(ws, struct workspace, list);
+	ZSTD_parameters params;
+	int size;
+
+	if (level > ZSTD_BTRFS_MAX_LEVEL)
+		level = ZSTD_BTRFS_MAX_LEVEL;
+
+	workspace->level = level > 0 ? level : ZSTD_BTRFS_DEFAULT_LEVEL;
```

Comment:
If we fail we shouldn't set the `workspace->level`. It doesn't matter currently, but it is better to not change the state if we fail.